validphys2/src/validphys/theorycovariance/construction.py

```diff
@@ -171,6 +171,37 @@ def combine_by_type(each_dataset_results_central_bytheory):
     )
     return process_info
 
+def process_starting_points(combine_by_type):
+    """Returns a dictionary of indices in the covariance matrix corresponding
+    to the starting point of each process."""
+    process_info = combine_by_type
+    running_index = 0
+    start_proc = defaultdict(list)
+    for name in process_info.theory:
+        size = len(process_info.theory[name][0])
+        start_proc[name] = running_index
+        running_index += size
+    return start_proc
+
+
+def covmap(combine_by_type, dataset_names):
+    """Creates a map between the covmat indices from matrices ordered by
+    process to matrices ordered by experiment as listed in the runcard"""
+    mapping = defaultdict(list)
+    start_exp = defaultdict(list)
+    process_info = combine_by_type
+    running_index = 0
+    for dataset in dataset_names:
+        size = process_info.sizes[dataset]
+        start_exp[dataset] = running_index
+        running_index += size
```

Comment:
I think this risks the same bug as before (I haven't tested it since I'm in the phone but I think it might be).

However, at this point (and also covmat by dataset) you always have the theory covmat so there's no need to use combine_by_type and simply hope for the best.

Rather you should look at the index of the theory covmat (as a dataframe) and select the index you want.

I would also argue that you don't need a mapping if you have a dataframe because you can simply ask for the first index of process X or dataset Y.

So the actual solution is, for the functions that use the theory (or experimental) covmat as an array, to use them as a dataframe.

If that's beyond the realm of possibility right now for whatever reason, this function should take also a covmat df as input (and output that same covmat together with the mapping) to avoid any ambigüities